### DISCUSSION TITLE
Dependency-dependant extension augmentation

### DISCUSSION THREAD

**Discussion by yrashk** [pgedc/pgedc#42] [Ideas]
Hey fellow @pgedc/members,

Wanted to explore something with you. I am building a constellation of extensions that know how to work with each other, but they are not always supposed to be strictly dependent. For example, I am building an authentication extension and I want to provide some building blocks for using it from HTTP endpoints facilitates by our HTTP server extension. 

But naturally, I don't want the authentication extension to _require_ the HTTP server.

I was thinking how should I approach it? One obvious answer is to create another extension (`auth_httpd`) that would require both `auth` and `httpd` extensions and provide whatever it is supposed to provide. Fine, but gets unwieldy and combinatorial :)

Another idea that I've worked with and seemed to have delivered good experience was to install event triggers and watch for the extension in question (such as `httpd`) to be installed and provision dependent parts upon creating this extension – while establishing a dependency for the cleanup. I was thinking of even generalizing this so that multiple extensions can easily reuse it without installing individual event triggers.

What do you think? How would you approach this?

**Comment by andreasscherbaum** [comment]
How do you plan to maintain the event trigger. Doesn't it come done to provide something similar to `auth_http`, which only provides the event trigger and the dependencies? Otherwise I don't see how you maintain these items separately and still make sure that they play well together.

**Reply by yrashk** [reply]
@andreasscherbaum, I am not sure what you mean. Can you elaborate?

I am inclined to generalize the problem over a singular event trigger that is called upon the creation of an extension that finds all functions adhering to specific naming/signature (let's say `when_extension_available(name)`) and calls them for every newly installed extension.

**Reply by andreasscherbaum** [reply]
Comes down to: where do you supply/store the configuration for the event trigger. Is this yet another extension?

**Reply by yrashk** [reply]
In the extensions themselves. As mentioned above, for example, by providing functions conforming to a certain naming and signature pattern.

**Reply by danolivo** [reply]
Based on our off-list discussion, I believe that creating a common DSM would be the best approach for a closed constellation of extensions. With this setup, you can create a Hash Table that has the entry format 'extname,funcname|funcptr'. This design allows extensions to publish their commonly used routines like an API.

In this scenario, another extension can check for the existence of registered routines, utilise a direct pointer when needed, or explore alternative methods to do the job. By implementing an update callback, the extension can receive notifications when a necessary extension is installed.

**Reply by yrashk** [reply]
Indeed, this would be useful for direct access to native functionality, as we discussed in person.

The general case of adaptive extensions would still be effectively catalog and event trigger based when direct access to the native component is not warranted.

**Reply by danolivo** [reply]
> The general case of adaptive extensions would still be effectively catalog and event trigger based when direct access to the native component is not warranted.

I may be wrong or misunderstand something entirely. How can we ensure that an extension's exported routine that hasn't been explicitly published as an API won't harm the extension's internal state?

**Reply by yrashk** [reply]
I think we may have been touching on different parts here. The extensions in questions have two API surfaces:

* Database ("SQL surface")
* Native C ABI

There are use-cases for both.

For example, the database surface is useful when an adaptive extension A can react to pre-existing or new appearance of extension B. As I was mentioning above, an extension may deploy an HTTP endpoint when omni_httpd is installed.

The Native C ABI surface is for the cases where related extensions need to relate – and I suspect it is the C ABI exporter's job to ensure that these entrypoints can't put into an invalid state. Happy to elaborate here.

**Reply by danolivo** [reply]
Alright, I understand. As I recall, we now have a system cache on the pg_extension. It is possible to register a callback to be triggered when this cache is updated and check the cache to see if there are changes. Is it an answer?

**Reply by yrashk** [reply]
Yes, it is possible, though the number of cache invalidation callbacks is somewhat limited – 64 total callbacks for syscache (`MAX_SYSCACHE_CALLBACKS`).

**Reply by danolivo** [reply]
> Yes, it is possible, though the number of cache invalidation callbacks is somewhat limited – 64 total callbacks for syscache (`MAX_SYSCACHE_CALLBACKS`).

I think Omnigres needs the only one core syscache callback ;) - constellation may use omni-extension of this call, as usual.

**Reply by yrashk** [reply]
Yes, that is the plan (I've done early experiments), where `omni` extension adds one hooks and can fan it out to others.

Longer term solution is to make that list resizable (the comments in the source code give this a nod), but the fan out is clearly available today.

**Comment by einhverfr** [comment]
So first it seems that the auth extension never should require the HTTP extension. The HTTP extension should be able to use the auth extension though.  An important side question is whether the presence of the shared libraries should allow the HTTP extension to use the auth functions or not.  That's a small technical detail and doesn't affect strategy but rather mechanism.  This may matter though based on the scope of the service (is it per database?  For the whole cluster?  For the whole server?)

My view is that the best way to handle this is with a runtime check on HTTP server startup and/or triggered by a reload etc.  I assume at the moment that this is a custom background worker (and that it doesn't listen on port 80 on *nix and so doesn't need root privileges).  I would then allocate a memory context as a child of TopMemoryContext for runtime checks and configuration.  Then I would make whatever runtime checks I wanted to do on startup and save the results there,    You could also reserve, for example, SIGHUP to rerun these checks.

Another option would be to have a list of runtime config options for the HTTP service where a trigger would check for relevant extensions on configuration.  An important problem here is that dropping the extension after configuring may have unexpected effects.  So you might still need a more limited event trigger.

At least that is how I would approach the problem.

**Reply by yrashk** [reply]
Perhaps I did a bit too much induction by giving a loaded auth+http example without getting into the details of it.

Let's simplify it to get to the essence of the postulated topic.

The idea here is that when extension B is installed, extension A should provision more objects (irrespective of whether B was installed before A or after), without B being a hard requirement for A. Let's say extension A defines function `do_something(int)` and extension B provides type `special_type`. We want to define `do_something` over `special_type`.

A straightforward approach is indeed to make user install a third extension C that requires A and B and defines `do_something` over `special_type`. Works for simple cases but requires higher cognitive load for the operator of the database (more extensions to provision)

The approach I am playing with is to enable extension A to do something (such as provisioning `do_something` over `special_type`) whenever B is installed – whether by having its own event trigger or by using a centralized one.

Does this clarify my original writing?

**Reply by andreasscherbaum** [reply]
> My view is that the best way to handle this is with a runtime check on HTTP server startup and/or triggered by a reload etc. I assume at the moment that this is a custom background worker (and that it doesn't listen on port 80 on *nix and so doesn't need root privileges). I would then allocate a memory context as a child of TopMemoryContext for runtime checks and configuration. Then I would make whatever runtime checks I wanted to do on startup and save the results there, You could also reserve, for example, SIGHUP to rerun these checks.

Without any kind of configuration this will lead to unpredictable results.

As example, depending on the presence of a library in a shared environment and enabling services based in that? This will enable the service even if no one actually wanted that service enabled.

**Comment by danolivo** [comment]
I have little experience designing constellations of extensions, but the idea with event trigger looks the best. As I remember, you connect all the extensions throughout the single connector, right? In that case, it makes sense to design a registration-based trigger subsystem that can help the extension identify the installation/drop of conflicting ones and, why not, execute some SQL scripts.
